Make identity the default value generation strategy

roji opened the pull request:
Closes #804

Notes:
* Setting the PG version to <10 in the context options changes the default back to serial.
* Disabled CI testing on 9.6 because we don't yet have a way to set the PG version in the context options on all tests (cc @austindrenski).

@ajcvickers @bricelam @AndriySvyryd, I have a specific question here for you guys... In a nutshell:
1. Pre-10 PG had a method for auto-increment columns called "serial". PG 10 introduced a much better one called "identity". The provider supported both methods, but had "serial" by default up to now. We want to take advantage of 3.0.0 to change the default behavior - that's what this PR does.
2. We already have a context option for compatibility mode: you can call SetPostgresVersion(), and that can affect SQL getting generated etc.
3. The convention which selects the default value generation strategy will now set a different default strategy on the model, based on this version context option. This way, users of PG 9.6 simply say "we're on 9.6" and they get the older serial columns. Same in scaffolding (NpgsqlAnnotationCodeGenerator) - the default depends on a context option.

I don't quite remember having a context option affecting setting/annotations on the *model* in this way. Do you see any problem with this?

AndriySvyryd commented:
@roji Should be fine. We do have the ProductVersion annotation that works similarly, though we don't expect users to set it.

roji commented:
Thanks @AndriySvyryd!

roji commented:
@austindrenski @YohDeadfall would appreciate a look on this.

This is mainly a risky change since it changes defaults for databases that already exist - we need to make sure everything works well and that users understand what's going on (docs). Here are some thoughts.

The difference between serial and identity (by default) is only in migrations. If I have an existing database and upgrade the provider, everything works as before. However, the next migration I do will attempt to migrate all my keys from serial to identity by default - which is a pretty big thing. I've done some testing and this seems to work fine, but some more eyes looking at this would be appreciated...

Of course, users can always opt out of this either by explicitly setting the strategy to serial, or by setting PG compatibility mode to 9.6 (which does the same implicitly).

YohDeadfall commented on src/EFCore.PG/Infrastructure/NpgsqlModelValidator.cs:
```diff
@@ -44,15 +44,19 @@ public override void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.
+                    $"'{strategy}' requires PostgreSQL 10.0 or later. " +
+                    "If you're using an older version, set PostgreSQL compatibility mode by calling " +
+                    $"'optionsBuilder.{nameof(NpgsqlDbContextOptionsBuilder.SetPostgresVersion)}()' in your model's OnConfiguring. " +
+                    "See the docs for more info");
```
nit: Lost dot at the end.

YohDeadfall commented on src/EFCore.PG/Design/Internal/NpgsqlAnnotationCodeGenerator.cs:
```diff
@@ -13,8 +15,16 @@ namespace Npgsql.EntityFrameworkCore.PostgreSQL.Design.Internal
+
+        public NpgsqlAnnotationCodeGenerator(
+            [NotNull] AnnotationCodeGeneratorDependencies dependencies,
+            [NotNull] IDbContextOptions options)
```
Is it possible to resolve `INpgsqlOptions` here?

roji replied:
Good question - I looked into this. Inside EF Core you have two provider options types: SqlServerOptions and SqlServerOptionsExtension. The former is only used as a singleton option, i.e. as a way to make sure that contexts with different options get different singletons. Actual settings which affect the different services are looked up in SqlServerOptionsExtension. I'm assuming this is primarily to allow for options which don't cause separate singletons (maybe @ajcvickers can shed some more light).

BTW we have some other places where services consume NpgsqlOptions instead of NpgsqlOptionsExtension, we can clean that up later.

roji replied:
Note - after some thought I ended up removing the version dependency for scaffolding. The idea is that we don't necessarily want the scaffolded database's version to actually impact or go into the scaffolded model, which may be used targeting a newer version etc.

Long story short, we now recognize both Serial and IdentityByDefault as by-convention strategies, which means they won't explicitly scaffolded into the model.

roji commented on doc/modeling/value-generation.md:
```diff
@@ -1,37 +1,46 @@
+* *Identity by default*: an identity column whose values are *by default* generated at the database, but you can still override this behavior by providing values from your application. This will generate the clause `GENERATED BY DEFAULT AS IDENTITY` on your column, and is the default value generation strategy.
 * *Identity always*: an identity column whose values are *always* generated at the database - you cannot provide values from your application. This will generate the clause `GENERATED ALWAYS AS IDENTITY` on your column.
-* *Identity by default*: an identity column whose values are *by default* generated at the database, but you can still override this behavior by providing values from your application. This will generate the clause `GENERATED BY DEFAULT AS IDENTITY` on your column.
+* *Serial*: the traditional PostgreSQL serial column. This will create the column with the `serial` datatype. Recommended only if you are using an older PostgreSQL version.
```
You mean smallint, int and bigint, right? The same goes for identity columns as well - I added a small sentence above.

roji commented on doc/modeling/value-generation.md:
```diff
@@ -1,37 +1,46 @@
 * *Sequence HiLo*: See below
 
-To maintain backwards compatibility with existing EF Core models, serial columns are still the default: when `ValueGeneratedOnAdd` is specified on a short, int or long property, the Npgsql EF Core provider will automatically map it to a `serial` column. Note that EF Core will automatically recognize key properties by convention (e.g. a property called `Id` in your entity) and will implicitly set  ...
+Since version 3.0.0, the Npgsql provider generates "identity by default" columns for ID columns. In other words, when `ValueGeneratedOnAdd` is specified on a short, int or long property, the Npgsql provider will automatically map it to an "identity by default" column. Note that EF Core will automatically recognize key properties by convention (e.g. a property called `Id` in your entity) and will  ...
```
You mean int/short/long? Seems like it would make the text full of these highlights and not very readable...

YohDeadfall replied:
It's a common practice to highlight types and we use it everywhere in the autogenerated docs. Why should this place be an exception?

roji replied:
OK, done.

roji commented on doc/modeling/value-generation.md:
```diff
@@ -1,37 +1,46 @@
+Since version 3.0.0, the Npgsql provider generates "identity by default" columns for ID columns. In other words, when `ValueGeneratedOnAdd` is specified on a short, int or long property, the Npgsql provider will automatically map it to an "identity by default" column. Note that EF Core will automatically recognize key properties by convention (e.g. a property called `Id` in your entity) and will  ...
+
+> [!CAUTION]
+> Changing a value generation is a significant change to an existing database. If you have an existing database with migrations, the first migration created after upgrading to version 3.0 will produce a migration that will alter your tables and convert serial columns to identity ones. This is a sensitive, one-time migration operation that should be done with care, and carefully tested before depl ...
```
Shouldn't it be the other way around? The long version should be in this page, with a "more information" link from the release notes...

roji commented on doc/release-notes/3.0.md:
```diff
@@ -11,3 +11,17 @@ The full list of issues for this release is [available here](https://github.com/
+> If you are using PostgreSQL 9.6 or earlier, it is recommended to call `SetPostgresVersion(9, 6)` in your model's `OnConfiguring` for general 9.6 compatibility,
+> which will include serial column. 
+>
+> More information is available in [the doc page about generated properties](../modeling/value-generation.md).
```
Yeah, makes sense.

roji commented on doc/toc.md:
```diff
@@ -14,6 +15,5 @@
 ## [Spatial Data](mapping/nts.md)
 ## [Full Text Search](mapping/full-text-search.md)
 ## [NodaTime](mapping/nodatime.md)
-# [Value Generation](value-generation.md)
```
Fixed

roji commented on src/EFCore.PG/Design/Internal/NpgsqlAnnotationCodeGenerator.cs:
```diff
@@ -174,5 +186,7 @@ public override MethodCallCodeFragment GenerateFluentApi(IIndex index, IAnnotati
             return null;
         }
+
+        bool PostgresVersionAtLeast(int major, int minor) => _postgresVersion is null || new Version(major, minor) <= _postgresVersion;
```
Refactored as an extension method over Version instead - an interface with a default implementation doesn't seem right for this.

That doesn't seem to work: `new Version(10, 1) < null` yields false, while `new Version(10, 1) > null` yields true. I guess Version's comparison operator treats null as the oldest possible version.

YohDeadfall replied:
Thought about an extension method, but interface default implementation allows to have less code (: The interface could be internal, so nobody would see this.

roji replied:
It's definitely an... original idea :) But I'd rather keep interfaces for polymorphism and extension methods seem to do the trick well enough...

roji commented:
Thanks for the review @YohDeadfall, addressed your comments.

In general no concerns with the general move that we're making, and possible bad consequences on users which I haven't thought of? @austindrenski would be good to get your thinking too.

YohDeadfall commented:
No concerns since users can opt-out I think.

YohDeadfall commented on src/EFCore.PG/Extensions/VersionExtensions.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        // Note: a null version is interpreted as the latest version and will always return true.
+        internal static bool AtLeast(this Version version, int major, int minor)
+            => version == null || version >= new Version(major, minor);
```
Oh, you have lost `is` which was previously (:

roji commented:
Scaffolding is broken.
